### ISSUE TITLE
Tackyons?

### ISSUE THREAD

**Issue by lovettchris** [leanprover/vscode-lean4#183]
> I missed this before, but so far we have mostly been using [Tachyons](http://tachyons.io/docs/table-of-styles/) rather than bespoke CSS classes for styling of this kind (with the exception of this exact file, of course). So instead of popup-menu-text one would write tc pa2 v-top dib as the className (pa1 or pa2, Tachyons intentionally doesn't provide specific pixel offsets for padding so as to keep a consistent, fixed scale). All the other classes either use var or need specific offsets to match a font, so it only applies to popup-menu-text. I will leave to you whether you want to change this, it's not the end of the world if we use classes.

Interesting.  Does VS Code itself actually use responsive UI design?  Meaning spacing/fontsize/padding change with screen size?

**Comment by lovettchris** [comment]
For example, this is a desktop screenshot followed by an iphone screenshot of the same project source file.  The iphone font is a bit different, but the spacing around everything is the same:

![image](https://user-images.githubusercontent.com/18707114/170401838-d1cae31d-fc89-4b54-a665-bdfb980ca772.png)

Interestingly LiveShare doesn't share "webviews" at all so InfoView is not even visible on an iPhone, but I also tried different resolution monitors from 1920x1080
to 1280x1024 and even 800x600 and the VS Code spacing around tooltips and everything else remained the same, here's the 800x600 monitor:

![image](https://user-images.githubusercontent.com/18707114/170402501-79437676-2cef-44f5-96e1-3eb4e001bf77.png)

**Comment by gebner** [comment]
> I missed this before, but so far we have mostly been using Tachyons rather than bespoke CSS classes

Basically we only have Tachyons because of widgets.  In Lean 3, widgets could send HTML snippets to the editor.  If these widgets could declare their own CSS classes and provide their own stylesheets, then there could be clashes if more than one widget was shown.  So Ed put in Tachyons as a standardized set of CSS classes.  Practically speaking, it also reduces the amount of data transferred, as `mr7` is shorter than `margin-right: 16rem`.

Personally, I strongly dislike these CSS wrappers (Tachyons, Tailwind, etc.).  Writing `class=mr7` is no improvement over `style='margin-right: 16rem'`.  On the contrary, it requires non-standard syntax and extra libraries.  And practically, it always seems to end up in copy&paste-programming where class name lists like `link pointer mh2 dim` end up being duplicated over and over again.  Just make a class, that's what they're there for!

I'd support removing Tachyons from all the components in the extension.

**Comment by lovettchris** [comment]
Having a standard set of CSS for widgets makes sense, I personally like the bootstrap library. I've been looking for good articles on "why use Tackyons" and the only thing I see find is that you want to do spacing in terms of variables like `.mv2 { margin-top:var(--spacing-small)}`, but if you are really doing responsive UI design, then you want different spacings per @screen size and the articles propose creating your own .mv2-a .mv2-b and so on for those different sizes at which point why do you need to use Tackyons at all?  So, I don't get it.   Sure you don't want to be styling HTML elements with `h1 {} and a{}` etc, except in a central "theme" (in this case inherited from VS code), and Tackyons is a small incremental step better than doing that. Perhaps that is all Tackyons is trying to be...they do pride themselves on being small and light weight... and I'm a fan of small and light weight to a point, there is a level of simplicity that becomes simplistic and it isn't really doing anything useful....and I see a lot of that in the nodejs world (notice we already have 323 nodejs packages!).

When I see `popup-menu-text` expanding to `tc pa2 v-top dib` it seems to me to be the opposite of object oriented programming, abstraction and encapsulation of complexity.  If I had to copy paste `tc pa2 v-top dib` 10 times and someone accidentally removed `dib` it would be much harder to notice that UI inconsistency creeping in.  I like abstract names that represent the "purpose" of that UI widget, namely, `popup-menu-text` and then the UI for that widget is controlled in one clear place.  

I guess what I wish I could do is something like `.popup-menu-text { tc pa2 v-top dib }` but that may require using a more advanced CSS preprocessor like [Sass](https://sass-lang.com/).  If I look at the CSS for this site https://microsoft.github.io/coyote/#concepts/actors/state-machine-demo/ it starts with a theme (windmill_dark) built on bootstrap, and then you can build custom widgets like the animation scrollbar which has it's own set of CSS properties all named like `player-control-*` which control different aspects of the look and feel and animation behavior of that widget and it all scales really nicely with responsive UI design, for example shrink the web browser down to a phone size, etc.  This is the way I'm used to using CSS (I built that site).  Am I missing some new trend in the industry here?

**Comment by Vtec234** [comment]
Ah, the functional CSS debate! Here is an article [for](https://adamwathan.me/css-utility-classes-and-separation-of-concerns/) and here is one [against](https://www.browserlondon.com/blog/2019/06/10/functional-css-perils/). The measured middle ground seems to be to abstract away repeated components but use `tc pa2 v-top dib` or whatever for components when that particular combination only appears once or twice. This seems to get the best of both approaches.

**Comment by lovettchris** [comment]
Thanks for the excellent articles, the first one answered my question about bundling utilities classes:

![image](https://user-images.githubusercontent.com/18707114/170779690-c2139163-2c36-428f-8566-f5ecf3b15d96.png)

Or using @apply I see in some examples.  I'd be fine with allowing that kind of abstraction, but trying hard to limit custom CSS properties, limiting ourselves to a standard set, like Tackyons, instead.  Of course my .codicon-blank class would still need to use explicit width and height properties.  So none of this helps with that kind of CSS.  

https://tailwindcss.com/ is interesting, Very cool animated highlighting of different html code in that page.  But the responsive screen size additions `sm:text-slate-500` etc, are a bit crazy.  Adam Wathan makes an interesting connection to abstraction by templates, which we do in React components, instead of abstraction in CSS.   

But I tend to agree more with functional-css-perils especially on the  problem of context sensitive styling, and theming.   We can't be using Tackyons like `.hover-black` and properly support VS code themes.  We'd probably want a version of Tackyons that deletes all fixed color classes.  There's quite a lot of them, probably 10% or more.

**Comment by lovettchris** [comment]
Meanwhile, vscode has been very responsive with this new feature, they added .codeicon-blank:
https://github.com/microsoft/vscode-codicons/issues/110
Now I have to figure out how to get the new version into our code so we can switch to this.

**Comment by EdAyers** [comment]
Hi I was the one who added tachyons. It's a divisive library. As Gabriel says it was mainly to save some headaches on the widgets stuff, however Lean 4 widgets are going to support injecting javascript directly so we are no longer necessarily dependent on it.

**Comment by lovettchris** [comment]
Ok, so I tend to agree with Gabriel then, especially if Widgets no longer needs it, I'm in favor of reducing complexity and removing stuff we no longer need.  Anyone object to removing our dependency on tachyons?

**Comment by Vtec234** [comment]
> I'm in favor of reducing complexity and removing stuff we no longer need. Anyone object to removing our dependency on tachyons?

Yes, sorry. We are using Tachyons all over the place, which means that to remove it you would need to write CSS classes to replace all the now-missing functionality. It's not clear whether such a change would reduce or increase complexity. It just seems like we have more important things to focus on at the moment - usability, robustness under various error conditions, widget features - than doing this kind of refactor. But it looks like I am in the minority, so if everyone else feels strongly that this should be done, I will not obstruct.

**Comment by EdAyers** [comment]
I really like tachyons tbc.

**Comment by lovettchris** [comment]
I agree with Wojciech this is not a major issue, but accumulating stuff we no longer need can lead to complex unmaintainable code, that's my only fear.  But as Wojciech says "We are using Tachyons all over the place" is true, I do see Tachyon class names mentioned in the current implementation of the Infoview and so it would be some work to remove that, and given this is low priority I propose we leave it as is for now and resolve this work item "leave them in for now", and revisit if we find a better reason to do that.

**Comment by lovettchris** [comment]
Thanks, yep, I was going to go there, but I'm glad you did instead.  I'm happy to do that cleanup, you are right it would be very quick and mechanical.  So, the next question is would this negatively impact the Widgets work?  Because I definitely don't want to hamper that effort in any way.  On the other hand, it is always good to ship "less dependencies" because then it is less work to clean it up later.  But Wojciech and Ed, can removing Tachyons be a zero hit on Widgets somehow?

**Comment by EdAyers** [comment]
It would make things harder for us if it was removed now, but we can go forward working with the assumption that the library will be gone one day and we should not depend on it.

**Comment by lovettchris** [comment]
Thanks, sounds like you would prefer we keep Tachyons in there.  My point about dependencies is that once 3rd party widgets become dependent on Tachyons being there, it will be impossible to remove it.  So the question is, how many Widgets will want to use it versus Widgets that want to use their own custom CSS that is not the same across widgets?  For example,  don't you already have some diverse Widget examples that use very different js/css libraries?  I would imagine that Tachyons is kind of niche enough that it won't be something that all Widget authors will be clamoring for.  Good platform design would hold off on taking that dependency until we start to see lots of requests from different Widget authors.  Bad platform design throws everything in just in case, and you end up with the kitchen sink.  Not saying we have anything bad yet, but it is something we need to constantly monitor.  One of my platform design guidelines is "less is more" and "only do what only you can do" in the platform, which keeps the platform tight and efficient.  Here I'm talking about the "lean4 vscode extension and vscode itself" is the platform on which "Widgets" will sit.  Lean4 itself of course is the real platform on which everything sits.  But in a way widgets are pure Lean, so the fact that depend on anything in vscode perhaps is a mistake, perhaps they should assume nothing about the vscode extension so they work equally well in all editors?

**Comment by Vtec234** [comment]
> Good platform design would hold off on taking that dependency until we start to see lots of requests from different Widget authors. Bad platform design throws everything in just in case, and you end up with the kitchen sink.

This is a great point. I agree we want to start off the user-widget API with a minimal API surface, and presumably that includes the CSS. I think it's okay if widgets take on a dependency on CSS classes that we use in the infoview, for example to make their fonts look consistent with it. But Tachyons has more than just these classes. The low cost solution would be to just say 'you can technically use this but don't depend on it existing'. This is already the case for infoview React components which are available to user-widgets but which may break at any time since the API is far from stable.

If we say this, then my only remaining objection to the Tachyons removal would just be 'unnecessary refactor'. But it's not a strongly held one, so I would approve a PR.

> But in a way widgets are pure Lean, so the fact that they depend on anything in vscode perhaps is a mistake, perhaps they should assume nothing about the vscode extension so they work equally well in all editors?

They explicitly do not :) The `@lean4/infoview` that widgets use is in this (vscode-lean4) monorepo for convenience, but it does not depend on VSCode. Widgets depend on e.g. React being present and React is loaded through the extension, true, but providing that is part of supporting the infoview in any editor.

**Comment by sarahsmith31** [comment]
We need to reach a consensus here at least for the Lean4 release. Sounds like removing Tachyons is the right thing to do to avoid excessive dependencies in the future. If these take off, it may reach a tipping point where removing them later would be monumental effort. What's required to remove them now successfully?
